File: SwaggerMan/Services/OpenAPIParser.swift
import Foundation
import OpenAPIKit30
import Yams

struct OpenAPIParser: OpenAPIParserProtocol {
    func parse(_ data: Data) throws -> ParsedSpec {
        if let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let swaggerVersion = dict["swagger"]
        {
            throw SwaggerManError.parsing(.unsupportedVersion(String(describing: swaggerVersion)))
        }

        if let doc30 = try? JSONDecoder().decode(OpenAPI.Document.self, from: data) {
            return buildSpec(from: doc30)
        }
        do {
            return try openAPIKit31ParseJSON(data)
        } catch {
            throw SwaggerManError.parsing(.invalidJSON(error.localizedDescription))
        }
    }

    func parseYAML(_ string: String) throws -> ParsedSpec {
        if let yamlAny = try? Yams.load(yaml: string) as? [String: Any],
           let swaggerVersion = yamlAny["swagger"]
        {
            throw SwaggerManError.parsing(.unsupportedVersion(String(describing: swaggerVersion)))
        }

        if let doc30 = try? YAMLDecoder().decode(OpenAPI.Document.self, from: string) {
            return buildSpec(from: doc30)
        }
        do {
            return try openAPIKit31ParseYAML(string)
        } catch {
            throw SwaggerManError.parsing(.invalidYAML(error.localizedDescription))
        }
    }

    // MARK: - Mapping

    private func buildSpec(from document: OpenAPI.Document) -> ParsedSpec {
        let info = SpecInfo(
            title: document.info.title,
            version: document.info.version,
            description: document.info.description
        )
        let servers = document.servers.map(\.urlTemplate.absoluteString)
        let operations = buildOperations(from: document)
        let securitySchemes = buildSecuritySchemes(from: document.components)
        return ParsedSpec(
            info: info,
            servers: servers,
            operations: operations,
            securitySchemes: securitySchemes.sorted { $0.name < $1.name },
            rawOperationCount: operations.count
        )
    }

    private func buildOperations(from document: OpenAPI.Document) -> [ParsedOperation] {
        var operations: [ParsedOperation] = []
        for (path, pathItemEither) in document.paths {
            guard let pathItem = resolvePathItem(pathItemEither, components: document.components) else {
                continue
            }
            let pathString = path.rawValue
            let pathLevelParams = pathItem.parameters
            for endpoint in pathItem.endpoints {
                guard let mappedMethod = mapMethod(endpoint.method) else { continue }
                let op = endpoint.operation
                let parsedParameters = (pathLevelParams + op.parameters).compactMap { paramEither -> ParsedParameter? in
                    guard let parameter = resolveParameter(paramEither, components: document.components) else {
                        return nil
                    }
                    return convertParameter(parameter, pathString: pathString)
                }
                let parsedRequestBody = op.requestBody.flatMap {
                    convertRequestBody($0, components: document.components)
                }
                operations.append(ParsedOperation(
                    id: "\(mappedMethod.rawValue) \(pathString)",
                    method: mappedMethod,
                    path: pathString,
                    operationId: op.operationId,
                    summary: op.summary,
                    description: op.description,
                    tags: op.tags ?? [],
                    parameters: parsedParameters,
                    requestBody: parsedRequestBody,
                    responses: buildResponses(op.responses, components: document.components)
                ))
            }
        }
        return operations
    }

    private func buildResponses(
        _ responses: OpenAPI.Response.Map,
        components: OpenAPI.Components
    ) -> [ParsedResponse] {
        var result: [ParsedResponse] = []
        for (statusCode, responseEither) in responses {
            guard let response = try? components.lookup(responseEither) else { continue }
            let schema: ParsedSchema? = response.content.first.flatMap { _, mediaType in
                guard let schemaEither = mediaType.schema else { return nil }
                switch schemaEither {
                case .a: return nil
                case let .b(jsonSchema): return convertSchema(jsonSchema)
                }
            }
            result.append(ParsedResponse(
                statusCode: statusCode.rawValue,
                description: response.description,
                schema: schema
            ))
        }
        return result.sorted { (Int($0.statusCode) ?? 999) < (Int($1.statusCode) ?? 999) }
    }

    private func buildSecuritySchemes(from components: OpenAPI.Components) -> [ParsedSecurityScheme] {
        components.securitySchemes.compactMap { entry in
            let kind = mapSecuritySchemeKind(entry.value.type)
            return ParsedSecurityScheme(
                id: entry.key.rawValue,
                name: entry.key.rawValue,
                kind: kind,
                description: entry.value.description
            )
        }
    }

    private func mapSecuritySchemeKind(_ type: OpenAPI.SecurityScheme.SecurityType) -> SecuritySchemeKind {
        switch type {
        case let .apiKey(name, location):
            let loc = switch location {
            case .header: "header"
            case .query: "query"
            case .cookie: "cookie"
            }
            return .apiKey(name: name, location: loc)
        case let .http(scheme, _):
            return .http(scheme: scheme)
        case .oauth2:
            return .oauth2
        default:
            return .unknown
        }
    }

    // MARK: - Helpers

    private func resolvePathItem(
        _ either: Either<JSONReference<OpenAPI.PathItem>, OpenAPI.PathItem>,
        components: OpenAPI.Components
    ) -> OpenAPI.PathItem? {
        switch either {
        case let .a(ref):
            components[ref]
        case let .b(item):
            item
        }
    }

    private func resolveParameter(
        _ either: Either<JSONReference<OpenAPI.Parameter>, OpenAPI.Parameter>,
        components: OpenAPI.Components
    ) -> OpenAPI.Parameter? {
        switch either {
        case let .a(ref):
            components[ref]
        case let .b(parameter):
            parameter
        }
    }

    private func mapMethod(_ method: OpenAPI.HttpMethod) -> HTTPMethod? {
        switch method {
        case .get: .get
        case .post: .post
        case .put: .put
        case .delete: .delete
        case .patch: .patch
        case .options: .options
        case .head: .head
        case .trace: nil
        }
    }

    private func mapLocation(_ location: OpenAPI.Parameter.Context.Location) -> ParameterLocation {
        switch location {
        case .query: .query
        case .header: .header
        case .path: .path
        case .cookie: .cookie
        }
    }

    private func convertParameter(_ parameter: OpenAPI.Parameter, pathString: String) -> ParsedParameter {
        let location = mapLocation(parameter.location)
        let resolvedSchema = parameter.schemaOrContent.a.flatMap {
            $0.schema.b ?? unwrapSchemaReference($0.schema)
        }
        let schema: ParsedSchema? = resolvedSchema.map { convertSchema($0) }

        return ParsedParameter(
            id: "\(pathString)-\(parameter.name)-\(location.rawValue)",
            name: parameter.name,
            location: location,
            required: parameter.required,
            schema: schema,
            description: parameter.description
        )
    }

    private func unwrapSchemaReference(_ either: Either<JSONReference<JSONSchema>, JSONSchema>) -> JSONSchema? {
        switch either {
        case .a:
            nil
        case let .b(schema):
            schema
        }
    }

    private func convertRequestBody(
        _ either: Either<JSONReference<OpenAPI.Request>, OpenAPI.Request>,
        components: OpenAPI.Components
    ) -> ParsedRequestBody? {
        let request: OpenAPI.Request
        switch either {
        case let .a(ref):
            guard let resolved = components[ref] else { return nil }
            request = resolved
        case let .b(value):
            request = value
        }

        guard let firstContent = request.content.first else {
            return ParsedRequestBody(required: request.required, contentType: "", schema: nil)
        }

        let contentType = firstContent.key.rawValue
        let schema: ParsedSchema? = if let schemaEither = firstContent.value.schema {
            switch schemaEither {
            case .a:
                nil
            case let .b(jsonSchema):
                convertSchema(jsonSchema)
            }
        } else {
            nil
        }

        return ParsedRequestBody(
            required: request.required,
            contentType: contentType,
            schema: schema
        )
    }

    private func convertSchema(_ schema: JSONSchema) -> ParsedSchema {
        let type: SchemaType
        var properties: [String: ParsedSchema]?
        var items: ParsedSchema?
        var required: [String]?

        switch schema.value {
        case .boolean:
            type = .boolean
        case .integer:
            type = .integer
        case .number:
            type = .number
        case .string:
            type = .string
        case let .object(_, objectContext):
            type = .object
            var props: [String: ParsedSchema] = [:]
            for (key, propSchema) in objectContext.properties {
                props[key] = convertSchema(propSchema)
            }
            properties = props
            required = objectContext.requiredProperties.isEmpty ? nil : objectContext.requiredProperties
        case let .array(_, arrayContext):
            type = .array
            if let itemSchema = arrayContext.items {
                items = convertSchema(itemSchema)
            }
        case .all, .one, .any, .not, .reference, .fragment:
            type = .unknown
        }

        let enumValues = schema.allowedValues?.compactMap { $0.value as? String }
        let defaultValue = schema.defaultValue.map { String(describing: $0.value) }
        let example = schema.example.map { String(describing: $0.value) }

        return ParsedSchema(
            type: type,
            properties: properties,
            items: items,
            enumValues: enumValues,
            required: required,
            defaultValue: defaultValue,
            example: example,
            description: schema.description
        )
    }
}

File: apps/desktop/src/core/openapi-parser.ts
// OpenAPI 3.0/3.1 (+ Swagger 2.0 일부) 문서를 ParsedSpec으로 변환.
// 로컬 $ref(#/components/...)를 해석한다. 외부 $ref는 미지원(첫 버전).
import yaml from "js-yaml";
import type {
  HTTPMethod,
  ParsedOperation,
  ParsedParameter,
  ParsedRequestBody,
  ParsedResponse,
  ParsedSchema,
  ParsedSecurityScheme,
  ParsedSpec,
  ParameterLocation,
  SchemaType,
} from "./types";
import { HTTP_METHODS } from "./types";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyObj = Record<string, any>;

export function parseSpecText(text: string): ParsedSpec {
  const trimmed = text.trim();
  let doc: AnyObj;
  if (trimmed.startsWith("{")) {
    doc = JSON.parse(trimmed);
  } else {
    doc = yaml.load(trimmed) as AnyObj;
  }
  if (!doc || typeof doc !== "object") {
    throw new Error("유효한 OpenAPI 문서가 아닙니다.");
  }
  return parseSpec(doc);
}

export function parseSpec(doc: AnyObj): ParsedSpec {
  const components: AnyObj = doc.components ?? {};

  const resolveRef = (node: AnyObj | undefined): AnyObj | undefined => {
    if (!node) return node;
    if (typeof node.$ref === "string") {
      const path = node.$ref.replace(/^#\//, "").split("/");
      let cur: AnyObj | undefined = doc;
      for (const key of path) {
        cur = cur?.[key];
      }
      return cur;
    }
    return node;
  };

  const toSchemaType = (raw: unknown): SchemaType => {
    const t = Array.isArray(raw) ? raw.find((x) => x !== "null") : raw;
    switch (t) {
      case "string":
      case "integer":
      case "number":
      case "boolean":
      case "array":
      case "object":
        return t;
      default:
        return "unknown";
    }
  };

  const convertSchema = (input: AnyObj | undefined, depth = 0): ParsedSchema | undefined => {
    if (!input || depth > 6) return undefined;
    const schema = resolveRef(input);
    if (!schema) return undefined;

    let type = toSchemaType(schema.type);
    if (type === "unknown" && schema.properties) type = "object";
    if (type === "unknown" && schema.items) type = "array";

    let properties: Record<string, ParsedSchema> | undefined;
    if (schema.properties && typeof schema.properties === "object") {
      properties = {};
      for (const [key, value] of Object.entries(schema.properties)) {
        const converted = convertSchema(value as AnyObj, depth + 1);
        if (converted) properties[key] = converted;
      }
    }

    return {
      type,
      properties,
      items: convertSchema(schema.items, depth + 1),
      enumValues: Array.isArray(schema.enum) ? schema.enum.map(String) : undefined,
      required: Array.isArray(schema.required) ? schema.required : undefined,
      defaultValue: schema.default != null ? String(schema.default) : undefined,
      example: schema.example != null ? String(schema.example) : undefined,
      description: schema.description,
    };
  };

  const buildParameters = (raw: unknown[] | undefined): ParsedParameter[] => {
    if (!Array.isArray(raw)) return [];
    const result: ParsedParameter[] = [];
    raw.forEach((item, index) => {
      const param = resolveRef(item as AnyObj);
      if (!param?.name || !param?.in) return;
      const location = param.in as ParameterLocation;
      if (!["path", "query", "header", "cookie"].includes(location)) return;
      result.push({
        id: `${location}-${param.name}-${index}`,
        name: param.name,
        location,
        required: param.required === true || location === "path",
        schema: convertSchema(param.schema),
        description: param.description,
      });
    });
    return result;
  };

  const buildRequestBody = (raw: AnyObj | undefined): ParsedRequestBody | undefined => {
    const body = resolveRef(raw);
    if (!body?.content || typeof body.content !== "object") return undefined;
    const entries = Object.entries(body.content as AnyObj);
    const preferred =
      entries.find(([ct]) => ct.includes("json")) ?? entries[0];
    if (!preferred) return undefined;
    const [contentType, mediaRaw] = preferred;
    const media = mediaRaw as AnyObj;
    // 스펙 example 우선순위: mediaType.example → mediaType.examples[*].value → schema.example
    let example: unknown = media?.example;
    if (example === undefined && media?.examples && typeof media.examples === "object") {
      const first = Object.values(media.examples as AnyObj)[0] as AnyObj | undefined;
      example = first?.value;
    }
    if (example === undefined) {
      const resolvedSchema = resolveRef(media?.schema);
      if (resolvedSchema?.example !== undefined) example = resolvedSchema.example;
    }
    return {
      required: body.required === true,
      contentType,
      schema: convertSchema(media?.schema),
      example,
    };
  };

  const buildResponses = (raw: AnyObj | undefined): ParsedResponse[] => {
    if (!raw || typeof raw !== "object") return [];
    const result: ParsedResponse[] = [];
    for (const [statusCode, value] of Object.entries(raw)) {
      const response = resolveRef(value as AnyObj);
      if (!response) continue;
      let schema: ParsedSchema | undefined;
      let example: unknown;
      if (response.content && typeof response.content === "object") {
        const first = Object.values(response.content as AnyObj)[0] as AnyObj | undefined;
        schema = convertSchema(first?.schema);
        // 응답 example: media.example > media.examples의 첫 value > 스키마 example
        if (first?.example !== undefined) {
          example = first.example;
        } else if (first?.examples && typeof first.examples === "object") {
          const firstEx = Object.values(first.examples as AnyObj)[0] as AnyObj | undefined;
          if (firstEx && "value" in firstEx) example = firstEx.value;
        }
        if (example === undefined) {
          const resolved = resolveRef(first?.schema);
          if (resolved?.example !== undefined) example = resolved.example;
        }
      }
      result.push({ statusCode, description: response.description, schema, example });
    }
    return result.sort(
      (a, b) => (parseInt(a.statusCode, 10) || 999) - (parseInt(b.statusCode, 10) || 999),
    );
  };

  const operations: ParsedOperation[] = [];
  let rawCount = 0;
  const paths: AnyObj = doc.paths ?? {};
  for (const [path, pathItemRaw] of Object.entries(paths)) {
    const pathItem = resolveRef(pathItemRaw as AnyObj);
    if (!pathItem) continue;
    const sharedParams = buildParameters(pathItem.parameters);
    for (const method of HTTP_METHODS) {
      const opRaw = pathItem[method.toLowerCase()] as AnyObj | undefined;
      if (!opRaw) continue;
      rawCount += 1;
      const opParams = buildParameters(opRaw.parameters);
      operations.push({
        id: `${method} ${path}`,
        method: method as HTTPMethod,
        path,
        operationId: opRaw.operationId,
        summary: opRaw.summary,
        description: opRaw.description,
        tags: Array.isArray(opRaw.tags) ? opRaw.tags : [],
        parameters: [...sharedParams, ...opParams],
        requestBody: buildRequestBody(opRaw.requestBody),
        responses: buildResponses(opRaw.responses),
      });
    }
  }

  const securitySchemes: ParsedSecurityScheme[] = [];
  const schemes: AnyObj = components.securitySchemes ?? {};
  for (const [name, value] of Object.entries(schemes)) {
    const scheme = resolveRef(value as AnyObj);
    if (!scheme) continue;
    switch (scheme.type) {
      case "apiKey":
        securitySchemes.push({
          name,
          kind: { kind: "apiKey", name: scheme.name ?? name, location: scheme.in ?? "header" },
        });
        break;
      case "http":
        securitySchemes.push({ name, kind: { kind: "http", scheme: scheme.scheme ?? "bearer" } });
        break;
      case "oauth2":
        securitySchemes.push({ name, kind: { kind: "oauth2" } });
        break;
      case "openIdConnect":
        securitySchemes.push({ name, kind: { kind: "openIdConnect" } });
        break;
      default:
        securitySchemes.push({ name, kind: { kind: "unknown" } });
    }
  }

  const servers: string[] = Array.isArray(doc.servers)
    ? doc.servers.map((s: AnyObj) => s.url).filter(Boolean)
    : doc.host
      ? [`${(doc.schemes?.[0] as string) ?? "https"}://${doc.host}${doc.basePath ?? ""}`]
      : [];

  return {
    info: {
      title: doc.info?.title ?? "Untitled API",
      version: doc.info?.version ?? "",
      description: doc.info?.description,
    },
    servers,
    operations,
    securitySchemes,
    rawOperationCount: rawCount,
  };
}
